Isolate a website into its own process when ITP informs of user interaction when shared process is enabled

rniwa opened the pull request:
#### 105b0b17d2379e30e2783968c57639d9baa4e089
<pre>
Isolate a website into its own process when ITP informs of user interaction when shared process is enabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=298805">https://bugs.webkit.org/show_bug.cgi?id=298805</a>
<a href="https://rdar.apple.com/160978534">rdar://160978534</a>

Reviewed by Ben Nham.

When ITP database has a record of user interactions with a given site, isolate the site into its own process
even when shared process is enabled for site isolation. To do this, we enumerate the set of all sites that
have had user interactions in WebsiteDataStore::fetchDomainsWithUserInteraction, and cache this result in
WebsiteDataStore. When there is a new user interaction with a registrable domain, we notify WebsiteDataStore
via WebProcessProxy::recordUserGestureAuthorizationToken, which updates the cache.

Test: TestWebKitAPI.SiteIsolation.SharedProcessWithResourceLoadStatistics
      TestWebKitAPI.SiteIsolation.SharedProcessAfterClick
      TestWebKitAPI.SiteIsolation.SharedProcessAfterKeyDown
      TestWebKitAPI.SiteIsolation.SharedProcessAfterUserInteractionInSharedProcesss

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
(WebKit::ResourceLoadStatisticsStore::loadWebsitesWithUserInteraction): Added. Returns a set of all sites
which has had user interactions in the past.

* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h:

* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::loadWebsitesWithUserInteraction): Added. Fetch the set of all sites
which has had user interactions using the background queue.

* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:

* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::fetchWebsitesWithUserInteractions): Added. Fetch the set of sites, in the form of
registrable domains, which has had user interactions in the past.

* Source/WebKit/NetworkProcess/NetworkProcess.h:

* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:

* Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
(WebKit::BrowsingContextGroup::sharedProcessForSite): Now takes a completion handler with FrameProcess
instead of synchronously returning it since now our decision may depend on an IPC to the network process.
(WebKit::BrowsingContextGroup::addFrameProcessAndInjectPageContextIf): Fixed a bug that this function was
trying to create remote pages even if the page had already been created. Keep track of the set of pages
that have already been created in the hash set (m_pagesInSharedProcess).

* Source/WebKit/UIProcess/BrowsingContextGroup.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::sendMouseEvent):
(WebKit::WebPageProxy::sendKeyEvent): Fixed a bug that we were using protectedLegacyMainFrameProcess() to
record the user gesture authorization token.
(WebKit::WebPageProxy::sendPreventableTouchEvent):
(WebKit::WebPageProxy::sendUnpreventableTouchEvent):
(WebKit::WebPageProxy::receivedNavigationActionPolicyDecision): Add the first party cookie access to the
shared process if we&apos;re loading the main frame.

* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::recordUserGestureAuthorizationToken): Inform the website data store that a given
site had user interactions.

* Source/WebKit/UIProcess/WebProcessProxy.h:

* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::fetchDomainsWithUserInteraction): Added.
(WebKit::WebsiteDataStore::didHaveUserInteractionForSiteIsolation): Added. Updates the set of registrable
domains that have had user interactions in the past. There is a subtle race condition to take care of in
this function. When a new site receives user interactions after fetchDomainsWithUserInteraction is called
(so that the set is being fetched from the network process) but before it receives the response from the
network process, the result we get from the network process might be missing this particular site. To avoid
this race condition, we store the relevant registrable domain in m_pendingDomainsWithUserInteractions under
this exact circumstances and add it to the set of sites we get back from the network process.

(WebKit::WebsiteDataStore::logUserInteraction):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::siteIsolatedViewWithSharedProcess):
(TestWebKitAPI::(SiteIsolation, SharedProcessWithResourceLoadStatistics)): Added.
(TestWebKitAPI::(SiteIsolation, SharedProcessAfterClick)): Added.
(TestWebKitAPI::(SiteIsolation, SharedProcessAfterKeyDown)): Added.
(TestWebKitAPI::(SiteIsolation, SharedProcessAfterUserInteractionInSharedProcesss)):

Canonical link: <a href="https://commits.webkit.org/300520@main">https://commits.webkit.org/300520@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a90f5de3711e6cffde0fd5323a3b3c50f93a6c58

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122802 "11 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42514 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33204 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129428 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/74906 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/25a7fa76-f278-4a95-97db-71e7d8b86276) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/124678 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43233 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51107 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93329 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61957 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f7a898d3-5034-4e57-bc12-0fc7fc72ec33) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125754 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/34560 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109929 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/74063 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/05cfc5ea-9b41-42ca-8ea1-7fa5a548e2ad) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33449 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28086 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72920 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104175 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28299 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132154 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/49747 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37869 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101841 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50124 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106145 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101716 "Found 3 new API test failures: WebKitGTK/TestInspector:/webkit/WebKitWebInspector/manual-attach-detach, TestWTF:CompletionHandlerDeathTest.MainThreadHandlerOnThreadAssertsDeathTest, TestWTF:CompletionHandlerDeathTest.ConstructionThreadHandlerOnThreadAssertsDeathTest (failure)") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47087 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25272 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/46518 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19391 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49604 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55357 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49071 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52423 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/50754 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
```diff
@@ -224,6 +224,24 @@ void WebResourceLoadStatisticsStore::populateMemoryStoreFromDisk(CompletionHandl
+    if (auto& domains = m_domainsWithUserInteraction)
+        return completionHandler(*domains);
+
+    postTask([this, proetectedThis = Ref { *this }, completionHandler = WTFMove(completionHandler)](auto& store) mutable {
```
Fixed.

rniwa commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -1616,6 +1616,22 @@ void NetworkProcess::setSessionIsControlledByAutomation(PAL::SessionID sessionID
 
+void NetworkProcess::fetchWebsitesWithUserInteractions(PAL::SessionID sessionID, CompletionHandler<void(const HashSet<RegistrableDomain>&)>&& completionHandler)
+{
+    static NeverDestroyed<HashSet<RegistrableDomain>> emptyHashSet;
```
Sure, we can do that.

rniwa commented on Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
```diff
@@ -45,23 +45,32 @@ BrowsingContextGroup::BrowsingContextGroup() = default;
+    websiteDataStore.fetchDomainsWithUserInteraction([protectedThis = Ref { *this }, websiteDataStore = Ref { websiteDataStore }, preferences = Ref { preferences }, site = Site { site },
+        lockdownMode = lockdownMode, pageConfiguration = Ref { pageConfiguration }, completionHandler = WTFMove(completionHandler)](const HashSet<WebCore::RegistrableDomain>& domainsWithUserInteraction) mutable {
+
+        if (domainsWithUserInteraction.contains(site.domain()))
```
Fixed with a test.

rniwa commented on Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
```diff
@@ -45,23 +45,32 @@ BrowsingContextGroup::BrowsingContextGroup() = default;
+        return completionHandler(nullptr);
     if (isMainFrame == IsMainFrame::Yes)
-        return nullptr;
+        return completionHandler(nullptr);
```
Fixed with a test.

rniwa commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
```diff
@@ -271,6 +272,8 @@ class WebResourceLoadStatisticsStore final : public ThreadSafeRefCountedAndCanMa
     WebCore::ResourceLoadStatistics::IsEphemeral m_isEphemeral { WebCore::ResourceLoadStatistics::IsEphemeral::No };
     HashSet<RegistrableDomain> m_domainsWithEphemeralUserInteraction;
 
+    std::optional<HashSet<RegistrableDomain>> m_domainsWithUserInteraction;
```
Removed the cache.

rniwa commented on Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
```diff
@@ -45,23 +45,32 @@ BrowsingContextGroup::BrowsingContextGroup() = default;
-    m_sharedProcess = frameProcess.ptr();
-    return frameProcess;
+        return completionHandler(nullptr);
+    websiteDataStore.fetchDomainsWithUserInteraction([protectedThis = Ref { *this }, websiteDataStore = Ref { websiteDataStore }, preferences = Ref { preferences }, site = Site { site },
```
But either way, there needs to be some mechanism for UI process to wait for network process since waiting for the network process to send us information is inherently racy with respect to the web content process triggering a load.

rniwa replied:
I've kept the current design for now. We can address it in a follow up.

rniwa commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -4008,7 +4008,7 @@ static std::optional<NativeWebMouseEvent> removeOldRedundantEvent(Deque<NativeWe
 {
     if (event.type() == WebEventType::MouseDown || event.type() == WebEventType::MouseUp)
-        processContainingFrame(frameID)->recordUserGestureAuthorizationToken(webPageIDInMainFrameProcess(), event.authorizationToken());
+        processContainingFrame(frameID)->recordUserGestureAuthorizationToken(frameID, webPageIDInMainFrameProcess(), event.authorizationToken());
```
Oh, that's a good point. I'd have to think about that more.

rniwa replied:
I've kept this code for now. We can address it in a follow up.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b40faef7e45e73b817396eb4e97c5c116f794da3. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51184

bnham commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -1466,8 +1466,13 @@ bool WebProcessProxy::wasPreviouslyApprovedFileURL(const URL& url) const
+void WebProcessProxy::recordUserGestureAuthorizationToken(FrameIdentifier frameID, PageIdentifier pageID, WTF::UUID authorizationToken)
 {
+    if (RefPtr dataStore = websiteDataStore()) {
+        if (RefPtr frame = WebFrameProxy::webFrame(frameID))
```
I think we only want to call `didHaveUserInteractionForSiteIsolation` if the frame is a main frame or if the target frame's site matches the main site right? Isn't the ITP heuristic "domains you've interacted with as a *first party* only"? We don't care about interactions with third party frames for this heuristic.

rniwa replied:
Oh, that's a good point. Asking...

bnham reviewed (approved):
Approved pending discussion with ITP folks on matching their heuristic.

rniwa commented on Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
```diff
@@ -224,6 +224,22 @@ void WebResourceLoadStatisticsStore::populateMemoryStoreFromDisk(CompletionHandl
+        HashSet<RegistrableDomain> domains;
+        if (RefPtr statisticsStore = store.m_statisticsStore)
+            domains = statisticsStore->loadWebsitesWithUserInteraction();
+        store.postTaskReply([domains = WTFMove(domains), completionHandler = WTFMove(completionHandler)] mutable {
```
sure, fixed.

rniwa committed:
Isolate a website into its own process when ITP informs of user interaction when shared process is enabled
https://bugs.webkit.org/show_bug.cgi?id=298805
rdar://160978534

Reviewed by Ben Nham.

When ITP database has a record of user interactions with a given site, isolate the site into its own process
even when shared process is enabled for site isolation. To do this, we enumerate the set of all sites that
have had user interactions in WebsiteDataStore::fetchDomainsWithUserInteraction, and cache this result in
WebsiteDataStore. When there is a new user interaction with a registrable domain, we notify WebsiteDataStore
via WebProcessProxy::recordUserGestureAuthorizationToken, which updates the cache.

Test: TestWebKitAPI.SiteIsolation.SharedProcessWithResourceLoadStatistics
      TestWebKitAPI.SiteIsolation.SharedProcessAfterClick
      TestWebKitAPI.SiteIsolation.SharedProcessAfterKeyDown
      TestWebKitAPI.SiteIsolation.SharedProcessAfterUserInteractionInSharedProcesss

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.cpp:
(WebKit::ResourceLoadStatisticsStore::loadWebsitesWithUserInteraction): Added. Returns a set of all sites
which has had user interactions in the past.

* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h:

* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::loadWebsitesWithUserInteraction): Added. Fetch the set of all sites
which has had user interactions using the background queue.

* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:

* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::fetchWebsitesWithUserInteractions): Added. Fetch the set of sites, in the form of
registrable domains, which has had user interactions in the past.

* Source/WebKit/NetworkProcess/NetworkProcess.h:

* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:

* Source/WebKit/UIProcess/BrowsingContextGroup.cpp:
(WebKit::BrowsingContextGroup::sharedProcessForSite): Now takes a completion handler with FrameProcess
instead of synchronously returning it since now our decision may depend on an IPC to the network process.
(WebKit::BrowsingContextGroup::addFrameProcessAndInjectPageContextIf): Fixed a bug that this function was
trying to create remote pages even if the page had already been created. Keep track of the set of pages
that have already been created in the hash set (m_pagesInSharedProcess).

* Source/WebKit/UIProcess/BrowsingContextGroup.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::sendMouseEvent):
(WebKit::WebPageProxy::sendKeyEvent): Fixed a bug that we were using protectedLegacyMainFrameProcess() to
record the user gesture authorization token.
(WebKit::WebPageProxy::sendPreventableTouchEvent):
(WebKit::WebPageProxy::sendUnpreventableTouchEvent):
(WebKit::WebPageProxy::receivedNavigationActionPolicyDecision): Add the first party cookie access to the
shared process if we're loading the main frame.

* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::recordUserGestureAuthorizationToken): Inform the website data store that a given
site had user interactions.

* Source/WebKit/UIProcess/WebProcessProxy.h:

* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::fetchDomainsWithUserInteraction): Added.
(WebKit::WebsiteDataStore::didHaveUserInteractionForSiteIsolation): Added. Updates the set of registrable
domains that have had user interactions in the past. There is a subtle race condition to take care of in
this function. When a new site receives user interactions after fetchDomainsWithUserInteraction is called
(so that the set is being fetched from the network process) but before it receives the response from the
network process, the result we get from the network process might be missing this particular site. To avoid
this race condition, we store the relevant registrable domain in m_pendingDomainsWithUserInteractions under
this exact circumstances and add it to the set of sites we get back from the network process.

(WebKit::WebsiteDataStore::logUserInteraction):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::siteIsolatedViewWithSharedProcess):
(TestWebKitAPI::(SiteIsolation, SharedProcessWithResourceLoadStatistics)): Added.
(TestWebKitAPI::(SiteIsolation, SharedProcessAfterClick)): Added.
(TestWebKitAPI::(SiteIsolation, SharedProcessAfterKeyDown)): Added.
(TestWebKitAPI::(SiteIsolation, SharedProcessAfterUserInteractionInSharedProcesss)):

Canonical link: https://commits.webkit.org/300520@main

webkit-commit-queue commented:
Committed 300520@main (105b0b17d237): <https://commits.webkit.org/300520@main>

Reviewed commits have been landed. Closing PR #51184 and removing active labels.
